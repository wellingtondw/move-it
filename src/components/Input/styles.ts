import styled, { css } from 'styled-components'
import * as ButtonStyles from 'components/Button/styles'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    height: 8rem;

    ${ButtonStyles.Button} {
      width: 8rem;
      border-radius: 0 ${theme.border.radius} ${theme.border.radius} 0;
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    width: 100%;
    background: linear-gradient(90deg, #4953B8 0%, rgba(73, 83, 184, 0.2) 100%);
    border-radius: ${theme.border.radius} 0 0 ${theme.border.radius};
    font-size: ${theme.font.sizes.small};
    padding-left: ${theme.spacings.small};
    color: ${theme.colors.secondaryText};
    border: none;
    outline: none;

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.secondaryText};
    }

    &::placeholder {
      color: ${theme.colors.secondaryText};
    }
  `}
`