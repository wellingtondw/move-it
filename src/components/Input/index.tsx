import * as S from './styles'

import Button from 'components/Button'
import { InputHTMLAttributes, useState } from 'react'

export type InputProps = {
  icon?: React.ReactNode
  initialValue?: string
  onInput?: (value: string) => void
} & InputHTMLAttributes<HTMLInputElement>

const Input = ({ icon, initialValue = '', onInput,...props }: InputProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)  
    
    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper>
      <S.Input {...props} onChange={onChange} value={value}/>
      {!!icon && (
        <Button icon={icon}/>
      )}
    </S.Wrapper>
  )
}

export default Input