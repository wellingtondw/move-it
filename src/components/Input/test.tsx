import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithTheme } from 'utils/tests/helpers'
import { RightArrowAlt } from '@styled-icons/boxicons-regular/RightArrowAlt'

import Input from '.'

describe('<Input />', () => {
  it('should render correctly', () => {
    const {container} = renderWithTheme(<Input placeholder='move it'/>)

    expect(screen.getByPlaceholderText('move it')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with button', () => {
    renderWithTheme(<Input  icon={<RightArrowAlt />}/>)

    expect(screen.getByRole('button')).toBeInTheDocument()
  })

  it('changes value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(<Input onInput={onInput} />)

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'

    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('Is accessible by tab', () => {
    renderWithTheme(<Input />)

    const input = screen.getByRole('textbox')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })
})
