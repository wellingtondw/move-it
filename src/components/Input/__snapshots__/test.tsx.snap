// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 8rem;
}

.c0 .sc-bdfBwQ {
  width: 8rem;
  border-radius: 0 0.5rem 0.5rem 0;
}

.c1 {
  width: 100%;
  background: linear-gradient(90deg,#4953B8 0%,rgba(73,83,184,0.2) 100%);
  border-radius: 0.5rem 0 0 0.5rem;
  font-size: 2.4rem;
  padding-left: 2.4rem;
  color: #B2B9FF;
  border: none;
  outline: none;
}

.c1:focus-within {
  box-shadow: 0 0 0.5rem #B2B9FF;
}

.c1::-webkit-input-placeholder {
  color: #B2B9FF;
}

.c1::-moz-placeholder {
  color: #B2B9FF;
}

.c1:-ms-input-placeholder {
  color: #B2B9FF;
}

.c1::placeholder {
  color: #B2B9FF;
}

<div
  class="c0"
>
  <input
    class="c1"
    placeholder="move it"
    value=""
  />
</div>
`;
